src/main/java/graphql/util/Traverser.java

```diff
@@ -83,67 +53,40 @@ private Traverser(TraverserState<T> traverserState, Function<? super T, ? extend
-     * @param initialData some data to passed into the traversal at the bootstrap time
-     *
-     * @return Traverser instance
-     */
     public static <T> Traverser<T> breadthFirst(Function<? super T, ? extends List<T>> getChildren, Object initialData) {
         return new Traverser<>(TraverserState.newQueueState(initialData), getChildren);
     }
 
 
-    /**
-     * Starts traversal of a tree from a provided root using specified Visitor
-     * and initial data to pass around Visitor's methods
-     *
-     * @param <U>     type of data argument to Visitor's methods
-     * @param root    subtree root to start traversal from
-     * @param visitor a Visitor object to be notified during traversal
-     *
-     * @return some data produced by the last Visitor's method invoked
-     */
-    public <U> Object traverse(T root, TraverserVisitor<? super T> visitor) {
+    public TraverserResult traverse(T root, TraverserVisitor<? super T> visitor) {
         return traverse(Collections.singleton(root), visitor);
     }
 
 
-    /**
-     * Starts traversal of a tree from a provided roots using specified Visitor
-     * and initial data to pass around Visitor's methods
-     *
-     * @param <U>     type of data argument to Visitor's methods
-     * @param roots   multiple subtree roots to start traversal from
-     *                can change that data to some other values of the same type or special Traverser
-     *                markers {@link TraversalControl}
-     * @param visitor a Visitor object to be notified during traversal
-     *
-     * @return some data produced by the last Visitor's method invoked
-     */
-    public <U> Object traverse(Collection<? extends T> roots, TraverserVisitor<? super T> visitor) {
+    public TraverserResult traverse(Collection<? extends T> roots, TraverserVisitor<? super T> visitor) {
         assertNotNull(roots);
         assertNotNull(visitor);
 
         traverserState.addNewContexts(roots, traverserState.newContext(null, null, rootVars));
 
         TraverserContext currentContext = null;
+        boolean cycle = false;
+        boolean fullTraversal = true;
         traverseLoop:
         while (!traverserState.isEmpty()) {
             Object top = traverserState.pop();
 
             if (top == TraverserState.Marker.END_LIST) {
                 // end-of-list marker, we are done recursing children,
                 // mark the current node as fully visited
-                TraversalControl traversalControl = visitor.leave((TraverserContext) traverserState.pop());
+                TraverserContext contextForLeave = (TraverserContext) traverserState.pop();
+                TraversalControl traversalControl = visitor.leave(contextForLeave);
                 assertNotNull(traversalControl, "result of leave must not be null");
+                assertTrue(CONTINUE_OR_QUIT.contains(traversalControl), "result can only return CONTINUE or QUIT");
```

Comment:
why not to put this assert into the default case? it would help to avoid lookup in the `CONTINUE_OR_QUIT` list.